validator/testcases/frameworks.py

```diff
@@ -0,0 +1,546 @@
+            if not self.CHECKED_FILES.search(path):
+                continue
+
+            identity = identified.get(path)
+            if not identity:
+                # This may turn out to be something we want:
+                # Skip checks on framework files even if not identified
+                # by hash.
+                # identified[path] = {"banned": True}
+
+                self.error(
+                    err_id=("testcases_frameworks",
+                            "unknown_file"),
+                    error="Unknown framework file",
+                    description=("A file belonging to the %s framework does "
+                                 "not match any known checksum." % self.NAME),
+                    filename=path)
+
+            elif not any(match[0:2] == tag for match in identity["sources"]):
+                version = (libraries.identify_files([identity], self.ID)
+                           or "unknown")
+
+                self.error(
+                    err_id=("testcases_frameworks",
+                            "incorrect_file_version"),
+                    error="Incorrect framework file version",
+                    description=("A file belonging to the %s framework does "
+                                 "not match the expected framework version."
+                                 % self.NAME,
+                                 "Expected %s, got %s."
+                                 % (self.version, version)),
+                    filename=path)
+
+        self.err.metadata["framework_files"] = self.framework_files
+
+        return self.valid
+
+    @classmethod
+    def detect_framework(cls, err, xpi):
+        for framework in cls.FRAMEWORKS:
+            f = framework(err, xpi)
+            if f.detect():
+                f.validate()
+                return f
+
+
+class BestToolbars(Framework):
+    NAME = "Besttoolbars"
+    METADATA_FILE = "chrome/content/config.json"
+
+    FRAMEWORK_FILES = ("chrome/content/ci.options.js",
+                       "chrome/content/framework.js",
+                       "chrome/content/subscriptloader.js",
+                       "chrome/content/config.json")
+
+    def detect(self):
+        FILE = "chrome/content/framework.js"
+        return (FILE in self.xpi and
+                "besttoolbars" in self.xpi.read(FILE))
+
+
+class Conduit(Framework):
+    NAME = "Conduit"
+    ACCEPTED = False
+
+    def detect(self):
+        return ("searchplugins/conduit.xml" in self.xpi and
+                any(f.startswith("components/Conduit")
+                    for f in self.xpi))
+
+
+class Crossrider(Framework):
+    NAME = "Crossrider"
+    ACCEPTED = False
+
+    def detect(self):
+        return any(f in self.xpi
+                   for f in (
+                       "chrome/content/CrossriderEXT.js",
+                       "chrome/content/crossrider.js",
+                       "chrome/content/crossriderapi.js"))
+
+
+class Jetpack(Framework):
```

Comment:
Add a docstring explaining that it's an "abstract" (or "base") framework, with `NAME = None`